### PR TITLE
Generalized PoW

### PR DESCRIPTION
**Pull request by paulhandy**
* Moved proof of work trait to curl crate

* Moved cpu proof of work implementation to curl-cpu crate

### CODE REVIEW

File: curl-cpu/src/cpupow.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                    return None;
+                }
+            }
+            Some(probe.trailing_zeros() as usize)
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (th0br0) [inline curl-cpu/src/cpupow.rs]: Could we move any PoW related tests to the `pow` crate similar to how we're doing tests with `curl`? That way, we don't need to duplicate tests for future pow impls.

File: curl-cpu/src/search.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+use tmath::*;
+use trytes::*;
+use alloc::Vec;
+use curl::*;
+use cpucurl::*;
+//use collections::String;
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: remove comment

File: curl-cpu/src/search.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        self[1].0 = 0b1111000111111000111111000111111000111111000111111000111111000111;
+        self[1].1 = 0b1000111111000111111000111111000111111000111111000111111000111111;
+        self[2].0 = 0b0111111111111111111000000000111111111111111111000000000111111111;
+        self[2].1 = 0b1111111111000000000111111111111111111000000000111111111111111111;
+        self[3].0 = 0b1111111111000000000000000000000000000111111111111111111111111111;
+        self[3].1 = 0b0000000000111111111111111111111111111111111111111111111111111111;
+    }
+}
+
+fn min(this: usize, that: usize) -> usize {
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: Use `core::cmp::min`?

File: curl-cpu/src/search.rs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+pub fn search_cpu<F>(input: Trinary, length: usize, group: usize, check: F) -> Option<Trinary>
+where
+    F: Fn(&[BCTrit]) -> Option<usize>,
+{
+    let mut curl = CpuCurl::<BCTrit>::default();
+    let bct: Vec<BCTrit> = input.trits();
+    curl.absorb(&bct); // should end before last absorb
+    let mut end = min(length, HASH_LENGTH);
+    (&mut curl.state[0..4]).offset();
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: why do this as a trait impl? just define an offset function? or will we need this in other parts again? can't remember

File: curl/src/lib.rs [hunk 5]
```diff
@@ -20,13 +22,24 @@ pub trait Sponge {
...
     T: Copy + Clone + Sized,
-
 {
     /// Absorb a `&[Trit]` into the sponge
     fn absorb(&mut self, trits: &[T]);
     /// Squeeze the sponge and return a `Vec<T>` with `trit_count` trits
     fn squeeze(&mut self, count: usize) -> Vec<T>;
 }
+
+pub trait PoWImpl {
```
Comment (th0br0) [inline curl/src/lib.rs]: let's just call this `pub trait ProofOfWork` ? It's not an `impl` as it's a `trait`.

File: curl/src/lib.rs [hunk 6]
```diff
@@ -20,13 +22,24 @@ pub trait Sponge {
...
+    /// Searches for a nonce given an `input` that gives a hash with `weight` zeros
+    /// Returns the nonce
+    fn search(input: Trinary, weight: u8) -> Option<Trinary>;
+}
+
+pub trait HamImpl {
```
Comment (th0br0) [inline curl/src/lib.rs]: as above. needs a better/ more detailed name?

File: pow/src/bct_search.rs [hunk 7]
```diff
@@ -0,0 +1,55 @@
+use tmath::*;
+use trytes::*;
+use search::Search;
+use curl::*;
+use alloc::Vec;
+use core;
+//use collections::String;
```
Comment (th0br0) [inline pow/src/bct_search.rs]: remove comment

File: curl/src/lib.rs [hunk 8]
```diff
@@ -20,13 +22,24 @@ pub trait Sponge {
...
+    /// Searches for a checksum given by hamming weight
+    /// Returns the nonce to create checksum
+    fn search(input: Trinary, length: u8, security: u8) -> Option<Trinary>;
```
Comment (th0br0) [inline curl/src/lib.rs]: maybe this should take a `&[Trit]` and we need to provide conversion traits to other trit encodings? mmh

File: tmath/src/increment.rs [hunk 9]
```diff
@@ -2,33 +2,35 @@ use trytes::*;
...
         for i in 0..self.len() {
             let (low, hi) = self[i];
             self[i].0 = hi ^ low;
             self[i].1 = low;
             if hi & !low == 0 {
-                break;
+                return self.len();
             }
         }
+        self.len() + 1
```
Comment (th0br0) [inline tmath/src/increment.rs:21]: why the return value/`+1`?

File: curl-cpu/src/search.rs (hunk 4 above)
Comment (paulhandy) [reply curl-cpu/src/search.rs]: It needs to be done for different word sizes

File: curl-cpu/src/cpupow.rs (hunk 1 above)
Comment (paulhandy) [reply curl-cpu/src/cpupow.rs]: I'll move it to the `curl` crate, since that's where the traits are.

File: tmath/src/increment.rs (hunk 9 above)
Comment (paulhandy) [reply tmath/src/increment.rs:21]: to notify that the "end" was greater than the previous end.

File: curl-cpu/src/search.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        self[1].0 = 0b1111000111111000111111000111111000111111000111111000111111000111;
+        self[1].1 = 0b1000111111000111111000111111000111111000111111000111111000111111;
+        self[2].0 = 0b0111111111111111111000000000111111111111111111000000000111111111;
+        self[2].1 = 0b1111111111000000000111111111111111111000000000111111111111111111;
+        self[3].0 = 0b1111111111000000000000000000000000000111111111111111111111111111;
+        self[3].1 = 0b0000000000111111111111111111111111111111111111111111111111111111;
+    }
+}
+
+pub fn search_cpu<F>(input: Trinary, length: usize, group: usize, check: F) -> Option<Trinary>
```
Comment (th0br0) [inline curl-cpu/src/search.rs]: could you document `length` and `group`?

File: tmath/src/increment.rs (hunk 9 above)
Comment (th0br0) [reply tmath/src/increment.rs:21]: which "end"? you never use the return value.

**Review by th0br0 — COMMENTED** [review]
added two questions that are still open for me, but looking good so far!

**Pull request closed without merging**